### PR TITLE
Refactor Governance data design

### PR DESCRIPTION
**Pull request by CostinCarabas**
Refactor Governance data design to have add downVetoVotes and abstain

### CODE REVIEW

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: energy-integration/governance-v2/src/views.rs [hunk 1]
```diff
@@ -44,11 +44,12 @@ pub trait ViewsModule:
...
     fn quorum_reached(&self, proposal_id: ProposalId) -> bool {
-        let total_votes = self.total_votes(proposal_id).get();
-        let total_downvotes = self.total_downvotes(proposal_id).get();
+        let proposal_votes = self.proposal_votes(proposal_id).get();
+        let total_up_votes = proposal_votes.up_votes;
+        let total_downvotes = proposal_votes.down_votes;
         let quorum = self.quorum().get();
 
-        &total_votes + &total_downvotes >= quorum && total_votes > total_downvotes
+        &total_up_votes + &total_downvotes >= quorum && total_up_votes > total_downvotes
```
[the remark this answers is not in the document]
Comment (CostinCarabas) [reply energy-integration/governance-v2/src/views.rs:52]: Agreed, will do that.

File: energy-integration/governance-v2/src/lib.rs [hunk 2]
```diff
@@ -147,8 +156,9 @@ pub trait GovernanceV2:
...
 
         let user_energy = self.get_energy_amount_non_zero(&voter);
-        self.total_votes(proposal_id)
-            .update(|total_votes| *total_votes += &user_energy);
+        self.proposal_votes(proposal_id).update(|proposal_votes| {
+            proposal_votes.up_votes += user_energy.clone();
+        });
 
         self.vote_cast_event(&voter, proposal_id, &user_energy);
     }
```
[the remark this answers is not in the document]
Comment (CostinCarabas) [reply energy-integration/governance-v2/src/lib.rs]: Below there are `fn down_vote_veto` and `fn abstain` but will aggregate them into one single function as agreed.

File: energy-integration/governance-v2/src/lib.rs [hunk 3]
```diff
@@ -147,31 +157,35 @@ pub trait GovernanceV2:
...
+                self.down_veto_vote_cast_event(&voter, proposal_id, &user_energy);
+            },
+            VoteType::AbstainVote => {
+                self.proposal_votes(proposal_id).update(|proposal_votes| {
+                    proposal_votes.abstain_votes += &user_energy.clone();
+                });
+                self.abstain_vote_cast_event(&voter, proposal_id, &user_energy);
+        
+            }
+            _ => sc_panic!(VOTE_OPTION_UNKOWN),
```
Comment (sasurobert) [inline energy-integration/governance-v2/src/lib.rs:187]: no need for this panic.

File: energy-integration/governance-v2/src/views.rs [hunk 4]
```diff
@@ -44,11 +44,16 @@ pub trait ViewsModule:
...
+        let proposal_votes = self.proposal_votes(proposal_id).get();
+        let total_votes = proposal_votes.get_total_votes();
+        let total_up_votes = proposal_votes.up_votes;
+        let total_down_votes = proposal_votes.down_votes;
+        let total_down_veto_votes = proposal_votes.down_veto_votes;
+        let third_total_votes = &total_votes / 3u64;
         let quorum = self.quorum().get();
 
-        &total_votes + &total_downvotes >= quorum && total_votes > total_downvotes
+
```
Comment (sasurobert) [inline energy-integration/governance-v2/src/views.rs:55]: rename the function to quorum_and_vote_reached as it does 2 things. Quorum check, like if there were at least quorum votes and checks whether the proposal_passed. maybe rename the function to proposal_passed.

File: energy-integration/governance-v2/src/lib.rs (hunk 3 above)
Comment (CostinCarabas) [reply energy-integration/governance-v2/src/lib.rs]: Done

File: energy-integration/governance-v2/src/views.rs (hunk 4 above)
Comment (CostinCarabas) [reply energy-integration/governance-v2/src/views.rs]: Done

File: energy-integration/governance-v2/tests/gov_test_setup/mod.rs [hunk 5]
```diff
@@ -149,14 +149,14 @@ where
...
+                sc.vote(proposal_id, VoteType::UpVote);
             })
     }
 
     pub fn downvote(&mut self, voter: &Address, proposal_id: usize) -> TxResult {
         self.b_mock
             .execute_tx(voter, &self.gov_wrapper, &rust_biguint!(0), |sc| {
-                sc.downvote(proposal_id);
+                sc.vote(proposal_id, VoteType::DownVote);
             })
```
Comment (sasurobert) [inline energy-integration/governance-v2/tests/gov_test_setup/mod.rs:160]: need new tests for noWithVeto, abstain, all the new logic.

**Review by sasurobert — APPROVED** [review]

**Pull request merged**